swev-id: django__django-14311 Fix autoreloader module detection

casey-brooks opened the pull request:
## Summary
- preserve the full module spec when the autoreloader restarts a process launched with `python -m module.path`
- only collapse to the parent package for `__main__.py` loaders so package execution keeps working
- add regression coverage for dotted specs, nested packages, and the bare `__main__` fallback

## Reproduction
1. Create a minimal package:
   ```
   mypkg/
     __init__.py
     start.py  # calls execute_from_command_line(sys.argv)
   ```
2. Start the server: `python -m mypkg.start runserver`
3. Modify a watched file to trigger reload.
4. Before this fix the autoreloader restarted with `python -m mypkg runserver` and crashed:
   ```
   Traceback (most recent call last):
     File ".../runpy.py", line NNN, in _run_module_as_main
       mod_name, mod_spec, code = _get_module_details(mod_name)
     File ".../runpy.py", line NNN, in _get_module_details
       raise ImportError(f"No module named {pkg}.__main__; '{pkg}' is a package and cannot be directly executed")
   ImportError: No module named mypkg.__main__; 'mypkg' is a package and cannot be directly executed
   ```
5. After this change Django restarts with `python -m mypkg.start runserver` so the autoreloader works for dotted module paths again.

## Upstream References
- Mirrors django/django@ec6d2531c59466924b645f314ac33f54470d7ac3 follow-up logic
- Aligns with django/django#13837

## Corner Cases Covered
- dotted but non-package modules keep their full spec name
- package `__main__.py` continues to restart via the package name
- nested module names and bare `__main__` fall back to the script path

## Testing
- `PYTHONPATH=$PWD DJANGO_SETTINGS_MODULE=tests.test_sqlite python tests/runtests.py utils_tests.test_autoreload --parallel=1`
- `flake8 django/utils/autoreload.py tests/utils_tests/test_autoreload.py`

casey-brooks committed:
fix(autoreload): preserve module spec on reload

Align get_child_arguments module handling with upstream logic.

Adds regression coverage for dotted specs, package __main__, nested modules, and bare __main__ fallback.

Fixes #317
Refs: django/django@ec6d253, PR #13837

casey-brooks commented:
Tests & Lint Summary:
- `PYTHONPATH=$PWD DJANGO_SETTINGS_MODULE=tests.test_sqlite python tests/runtests.py utils_tests.test_autoreload --parallel=1` (83 passed, 0 failed, 20 skipped)
- `flake8 django/utils/autoreload.py tests/utils_tests/test_autoreload.py` (no issues)

noa-lucent commented on django/utils/autoreload.py:
```diff
@@ -223,9 +223,20 @@ def get_child_arguments():
+            args += ['-m', module_name]
+            args += sys.argv[1:]
+            return args
     elif not py_script.exists():
```
[major] When `__spec__` is present but we don't derive a module name (e.g. `spec.name == '__main__'`), this new structure skips the Windows shim branch entirely because the `elif not py_script.exists()` only runs when `spec` is `None`. Previously, those cases still hit the fallback because the first condition required a truthy `spec.parent`. That branch is what repairs console-script launches where `sys.argv[0]` lacks the `.exe`/`-script.py` suffix. Dropping it means we now re-run `[sys.executable, sys.argv[0], …]` and can no longer find the real wrapper. Could we turn this into an unconditional `if not py_script.exists():` after the spec block so the compatibility path still runs when `module_name` is false?

casey-brooks replied:
Thanks for the catch — I've dropped the `elif` chain and now run the Windows entrypoint shim any time the script path is missing, even when `__spec__` exists but doesn't yield a module name. Added coverage for both the bare `__main__` script case and the spec-without-module exe shim to make sure the shim stays wired up.

noa-lucent reviewed (changes requested):
The spec-handling looks good overall, but we need to keep the Windows entrypoint fallback wired up for cases where  is populated but no module name is computed. I've left details inline.

casey-brooks committed:
fix(autoreload): run windows shim when spec empty

Ensure get_child_arguments reuses Windows entrypoint fallback even when __spec__ lacks a usable module name.

Add regression coverage for bare __main__ scripts with a real file and for spec-only runs that rely on exe shims.

Refs: #317

casey-brooks commented:
Updated Tests & Lint:
- `PYTHONPATH=$PWD DJANGO_SETTINGS_MODULE=tests.test_sqlite python tests/runtests.py utils_tests.test_autoreload --parallel=1` (84 passed, 0 failed, 20 skipped)
- `flake8 django/utils/autoreload.py tests/utils_tests/test_autoreload.py` (no issues)

noa-lucent reviewed (approved):
Windows shim now runs regardless of  presence, and the regression tests cover both dotted module and shim scenarios. Looks good.
